### PR TITLE
[HTTP-38] - Fixing issue with handling Http timeout where Event task was not removed from Flink's AsyncIO operator internal queue.

### PR DESCRIPTION
**Pull request by kristoffSC**
Fixes https://github.com/getindata/flink-http-connector/issues/38

The Async version of LookupSource is guarded by two timeout timers.

First one is coming from Flink's AsyncIO operator with default value set to 3 minutes.
The second one is set by connect code on HTTP request and currently is set to 2 minutes.

In case of HTTP request timeout, the the task representing this request is not removed from Flink's AsyncIO intenral task queue.
This will eventually fill the entire queue (default value 100) and created backpresure on upstream elements.

In case of any timeout or exception coming from Http client, task should be removed from Flink's queue.

In this PR we make sure that for case when we had an Exception from Http Client, we will call `resultFuture.complete` on `CompletableFuture` commign from Flink Runtime in `AsyncHttpTableLookupFunction.eval` method.

this PR also adds a custom thread pools for Http Sink calls, where previously JVM shared ForkJoinPool was used.

At the end, various settings for HTTP and Flink AsyncIO timeouts were extracted to properties.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -150,9 +150,20 @@ This can be done by using option like so: `'lookup-request.format.customFormatNa
...
 
-In configuration when default, Flink-Json format is used for `GenericGetQueryCreator`, all options defined in [json-format](https://nightlies.apache.org/flink/flink-docs-master/docs/connectors/table/formats/json/)
+In configuration with default, Flink-Json format is used for `GenericGetQueryCreator`, all options defined in [json-format](https://nightlies.apache.org/flink/flink-docs-master/docs/connectors/table/formats/json/)
 can be passed through table DDL. For example `'lookup-request.format.json.fail-on-missing-field' = 'true'`. In this case, format identifier is `json`.
 
+#### Timeouts
+Lookup Source is guarded by two timeout timers. First one is specified by Flink's AsyncIO operator that executes `AsyncTableFunction`.
+The default value of this timer is set to 3 minutes and can be changed via `table.exec.async-lookup.timeout` [option](https://nightlies.apache.org/flink/flink-docs-master/docs/dev/table/config/#table-exec-async-lookup-timeout).
+
+The second one is set per individual HTTP requests by HTTP client. It's default value is set currently to 30 seconds and can be chane via `gid.connector.http.source.lookup.request.timeout` option.
```
Comment (grzegorz8) [inline README.md:160]: nit:

- `It's` -> Its
- chane -> changed

File: README.md [hunk 2]
```diff
@@ -150,9 +150,20 @@ This can be done by using option like so: `'lookup-request.format.customFormatNa
...
 can be passed through table DDL. For example `'lookup-request.format.json.fail-on-missing-field' = 'true'`. In this case, format identifier is `json`.
 
+#### Timeouts
+Lookup Source is guarded by two timeout timers. First one is specified by Flink's AsyncIO operator that executes `AsyncTableFunction`.
+The default value of this timer is set to 3 minutes and can be changed via `table.exec.async-lookup.timeout` [option](https://nightlies.apache.org/flink/flink-docs-master/docs/dev/table/config/#table-exec-async-lookup-timeout).
+
+The second one is set per individual HTTP requests by HTTP client. It's default value is set currently to 30 seconds and can be chane via `gid.connector.http.source.lookup.request.timeout` option. 
+
+Flink's current implementation of `AsyncTableFunction` does not allow specifying custom logic for handling Flink AsyncIO timeouts as it is for Java API.
+Because of that, if AsyncIO timer pass, Flink will throw TimeoutException which will case job restart.
```
Comment (grzegorz8) [inline README.md:163]: nit: `pass` -> `passes`, `case` -> `cause`?

File: README.md [hunk 3]
```diff
@@ -150,9 +150,20 @@ This can be done by using option like so: `'lookup-request.format.customFormatNa
...
+
+The HTTP request timeouts on the other hand will not cause Job restart. In that case, proper log information will be printed application logs only.
```
Comment (grzegorz8) [inline README.md:165]: Some preposition is missing in this sentence `log information will be printed application logs only.`

File: src/main/java/com/getindata/connectors/http/internal/security/SecurityContext.java [hunk 4]
```diff
@@ -24,6 +24,8 @@
 
 import lombok.extern.slf4j.Slf4j;
 
+import com.getindata.connectors.http.internal.utils.ConfigUtils;
```
Comment (grzegorz8) [inline src/main/java/com/getindata/connectors/http/internal/security/SecurityContext.java:27]: nit: use static import.

File: src/main/java/com/getindata/connectors/http/internal/utils/JavaNetHttpClientFactory.java [hunk 5]
```diff
@@ -22,8 +23,56 @@
 @NoArgsConstructor(access = AccessLevel.NONE)
 public class JavaNetHttpClientFactory {
 
+    /**
+     * Creates Java's {@link HttpClient} instance that will be using default,JVM shared {@link
+     * java.util.concurrent.ForkJoinPool} for async cals. calls.
```
Comment (grzegorz8) [inline src/main/java/com/getindata/connectors/http/internal/utils/JavaNetHttpClientFactory.java:28]: `for async cals. calls.` ?

File: src/main/java/com/getindata/connectors/http/internal/security/SecurityContext.java (hunk 4 above)
Comment (kristoffSC) [reply src/main/java/com/getindata/connectors/http/internal/security/SecurityContext.java:27]: I did not use static import on purpose here.

There is debate, whether static imports are "ok" or not, and the con argument is that  from code you dont see that this const or method is not defined in this class, and you need to scroll up to find the import etc. Still nothing that IDE can sort out, but when reading code on GitHub -> this information is missing.

So this is something I'm trying here now, although I'm still using static imports for `assertThat` and all that jazz for Unit tests.

**Review by swtwsk — APPROVED** [review]
Apart from the comments that @grzegorz8 made, looks good to me :)

File: README.md (hunk 1 above)
Comment (kristoffSC) [reply README.md:160]: done

File: README.md (hunk 2 above)
Comment (kristoffSC) [reply README.md:163]: done

File: README.md (hunk 3 above)
Comment (kristoffSC) [reply README.md:165]: fixed

**Pull request merged**